File: examples/example-55/js/features/utils/http.js
const axios = require('axios');
const config = require('config');

async function executeHttpRequestInternal({
    url, method = 'GET', headers = {}, body = null }) {
    const response = await axios({
        method,
        url,
        headers,
        data: body
    });
    return response;
}

async function executeHttpRequest(world, request) {
    let res;
    world.http = world?.http ?? {};
    world.http.requests = world?.http?.requests ?? [];
    world.http.responses = world?.http?.responses ?? [];
    const ignoreErrors = world.http.ignoreErrors || config?.http?.ignoreErrors;
    try {
        const req = { ...request, url: `${config?.http?.baseUrl}${request.url}` };
        world.http.requests.unshift(req);
        res = await executeHttpRequestInternal(req);
    } catch (err) {
        if (!err.response || !ignoreErrors) {
            throw err;
        }
        res = err.response;
    } finally {
        world.http.responses.unshift(res);
    }
    return res;
}

module.exports = { executeHttpRequest };

File: examples/example-56/ruby/features/support/http.rb
require 'uri'
require 'net/http'
require 'json'
require './features/utils/HttpTest'

def execute_http_request_internal(url:, method: 'GET', headers: {}, body: nil)
    uri = URI.parse(url)
    http = Net::HTTP.new(uri.host, uri.port)
    http.use_ssl = uri.scheme == 'https'

    request_class = Net::HTTP.const_get(method.capitalize)
    request = request_class.new(uri)

    request.body = body.to_json if %w[POST PUT PATCH].include?(method.upcase) && body

    headers.each { |key, value| request[key] = value }

    http.request(request)
rescue NameError
    raise "Unsupported HTTP method: #{method}"
end

def execute_http_request(request)
    url = "#{Settings.http.baseUrl}#{request[:url]}"
    ignore_errors = HttpTest.get_ignore_errors || Settings.http.ignoreErrors
    HttpTest.add_request(request)
    res = execute_http_request_internal(
        url: url,
        method: request[:method],
        headers: request[:headers],
        body: request[:body]
    )
ensure
    HttpTest.add_response(res)
    raise "HTTP Error: #{res.code} #{res.message}" if !ignore_errors &&
                                                      res.code.to_i >= 400

    res
end

Before('@http-ignore-errors') do
    HttpTest.set_ignore_errors(true)
end

After do
    HttpTest.reset
end
